### ISSUE TITLE
Consider `Span` to be an `associatedType` on `Tracer`

### ISSUE THREAD

**Issue by fabianfett** [apple/swift-distributed-tracing#78]
Currently the `Tracer` protocol has the `startSpan` requirement. The return of `startSpan` is a `Span` which is a protocol. In other words we return an existential here. Instead we might want to consider making the Span an associated type and returning the concrete type:

```swift
public protocol Tracer: Instrument {
    associatedType Span: SpanProtocol // the current Span protocol maybe needs a rename to SpanProtocol for this?

    func startSpan(
        _ operationName: String,
        baggage: Baggage,
        ofKind kind: SpanKind,
        at time: DispatchWallTime,
        function: String,
        file fileID: String,
        line: UInt
    ) -> Span

    func forceFlush()
}
```

I can see no reasons, why a Tracer would return Spans of different types. If that was a requirement adopters could work around this, by using internal enums.

### Result

We preserve more type information if we work with a concrete Tracer.

**Comment by slashmo** [comment]
Sounds good to me 👍 The rename to `SpanProtocol` also sounds good and actually fits nicely with @ktoso's rename of `Tracer` and `Instrument` to `*Protocol`: #69

**Comment by ktoso** [comment]
Yeah, sounds like a good idea -- it does seem like we'll indeed adopt more `...Protocol` types, so this sounds good as well.

**Comment by ktoso** [comment]
Done, I'll do one more rename though -- we'll keep `protocol Span`
